From: Karsten Keil <kkeil@novell.com>
Subject: Fix bad eeprom checksum misdetection on some notebooks
Reference: 190417 235315
Mainline: never

This workaround fix a problem on some Lenovo notebooks where the first
call to e1000_validate_eeprom_checksum fails, because the PCIe was in a deep
powersave state. Better would be that Lenovo fix this via a BIOS update, but
maybe that will never happen...

---
 drivers/net/e1000/e1000_main.c |    8 ++++++--
 1 files changed, 6 insertions(+), 2 deletions(-)

--- linux-2.6.20.orig/drivers/net/e1000/e1000_main.c
+++ linux-2.6.20/drivers/net/e1000/e1000_main.c
@@ -1018,8 +1018,12 @@ e1000_probe(struct pci_dev *pdev,
 	/* make sure the EEPROM is good */
 
 	if (e1000_validate_eeprom_checksum(&adapter->hw) < 0) {
-		DPRINTK(PROBE, ERR, "The EEPROM Checksum Is Not Valid\n");
-		goto err_eeprom;
+		/* On some hardware the first attemp fails */
+		if (e1000_validate_eeprom_checksum(&adapter->hw) < 0) {
+			DPRINTK(PROBE, ERR, "The EEPROM Checksum Is Not Valid\n");
+			goto err_eeprom;
+		} else
+			DPRINTK(PROBE, INFO, "The EEPROM Checksum failed in the first read, now OK\n");
 	}
 
 	/* copy the MAC address out of the EEPROM */
